File: R-pkg/rpm/R/matrixops.R

rowwise.sums <- function(mat, vec) {
  stopifnot(ncol(mat) == length(vec))

  result <- matrix(0.0, nrow=nrow(mat), ncol=ncol(mat))
  result <- .C("rowwise_sums",
               as.numeric(mat),
               as.numeric(vec),
               as.integer(nrow(mat)), as.integer(ncol(mat)),
               result=result, PACKAGE='rpm')$result
  result
}

File: R-pkg/rpm/src/matrixops.c
#include "rpm.h"

/**
 * mat: m x n matrix
 * vec: n-length vector
 * result: m x n matrix
 */
void rowwise_sums(double *_mat, 
		  double *vec,
		  int *_m,
		  int *_n,
		  double *_result) {
  const int m = *_m;
  const int n = *_n;

  Matrix mat = {_mat, m, n};
  Matrix result = {_result, m, n};

  for (int i=0;i<m;i++) {
    for (int j=0;j<n;j++) {
      *get(&result, i, j) = *get(&mat, i, j) + vec[j];
    }
  }
}
